### PR TITLE
CSS143B HW2 

### PR DESCRIPTION
**Pull request by ApologytoChrist**

### CODE REVIEW

File: src/main/java/Problem2.java [hunk 1]
```diff
@@ -7,6 +7,22 @@
...
+        for (int i = 1; i < nums.length; i++) {
+            number = nums[i];
+            prevNumber = nums[i-1];
+
+            if (number != prevNumber) {
+                nums[difNumIndex] = number;
+                difNumIndex++;
+            }
+        }
+        for (int i = difNumIndex; i < nums.length; i++) {
```
Comment (jmkelly6) [inline src/main/java/Problem2.java]: Wrong algorithm. Only one loop is used in O(1) complexity

File: src/test/java/Problem2Test.java [hunk 2]
```diff
@@ -7,13 +7,23 @@
...
         int[][] inputs = {
                 // DO NOT REMOVE OR CHANGE THE GIVEN TEST CASES!
-                {0, 0, 1, 1, 1, 2, 2, 4, 4, 6}
+                {0, 0, 1, 1, 1, 2, 2, 4, 4, 6},
                 // add more test cases
+                {0, 0, 0, 0, 0, 0},
+                {0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
+                {0, 1, 1, 1, 1, 2, 4, 5, 8, 8, 8, 8, 8, 8, 9, 9, 9},
+                {-2, -2, -1, -1, -1 ,3, 6, 6, 6, 6, 6, 9, 9, 12, 100, 100},
+                {7, 7, 7, 7, 14, 14, 41, 41, 41,41, 99, 99}
```
Comment (jmkelly6) [inline src/test/java/Problem2Test.java:16]: Recommendation for future assignments: check edge cases (ex. if an object is NULL or empty).

File: src/main/java/Problem3/Book.java [hunk 3]
```diff
@@ -15,11 +15,17 @@ public Book(String title, String author) {
...
-        // homework
+        this.title = anotherBook.title;
+        this.author = anotherBook.author;
+        this.id = anotherBook.id;
     }
 
     @Override
     public boolean equals(Object obj) {
-        // homework
+        if(this.getClass() == obj.getClass()) {
```
Comment (jmkelly6) [inline src/main/java/Problem3/Book.java:25]:
```suggestion
        if (obj == this) {
            return true;
        }
```

File: src/main/java/Problem3/Movie.java [hunk 4]
```diff
@@ -9,15 +9,24 @@
...
         // homework
+        this.title = anotherMovie.title;
+        this.rating = anotherMovie.rating;
+        this.id = anotherMovie.id;
     }
 
     @Override
     public boolean equals(Object obj) {
         // homework
+        if(this.getClass() == obj.getClass()) {
```
Comment (jmkelly6) [inline src/main/java/Problem3/Movie.java:26]: See Book.java for similar suggestion

File: src/main/java/Problem3/MovieAction.java [hunk 5]
```diff
@@ -21,6 +19,13 @@ public int getLateFeeInDollar() {
 
     @Override
     public int calcLateFees(int numOfDaysPastDue) {
-        // homework
+        if (numOfDaysPastDue <= 0 ) {
+           return 0;
+        }
+        else if (numOfDaysPastDue >= 5) {
```
Comment (jmkelly6) [inline src/main/java/Problem3/MovieAction.java]:
```suggestion
        int fee = lateFeePerDayInDollar * numOfDaysPastDue;
        return (numOfDaysPastDue >= 5) ? 2 * fee : fee;
```

File: src/main/java/Problem2.java (hunk 1 above)
Comment (pdgetrf) [reply src/main/java/Problem2.java]: this loop is not needed.

File: src/main/java/Problem3/BookFiction.java [hunk 6]
```diff
@@ -24,6 +24,10 @@ public int getLateFeeInDollar() {
 
     @Override
     public int calcLateFees(int numOfDaysPastDue) {
-        // homework
+        if (numOfDaysPastDue <= 0 ) {
+            return 0;
+        } else {
```
Comment (pdgetrf) [inline src/main/java/Problem3/BookFiction.java:29]: "else" is not needed

File: src/main/java/Problem3/BookRomance.java [hunk 7]
```diff
@@ -21,6 +19,10 @@ public int getLateFeeInDollar() {
 
     @Override
     public int calcLateFees(int numOfDaysPastDue) {
-        // homework
+        if (numOfDaysPastDue <= 0 ) {
+            return 0;
+        } else {
```
Comment (pdgetrf) [inline src/main/java/Problem3/BookRomance.java:24]: "else" is not needed here

**Review by pdgetrf — APPROVED** [review]
